File: backend/routes/documents.py
"""
routes/documents.py
───────────────────
Endpoints for document upload, listing, and metadata retrieval.
"""
from typing import Annotated, List
from fastapi import APIRouter, Depends, HTTPException, status, UploadFile, File
from fastapi.responses import StreamingResponse
from io import BytesIO
from sqlalchemy.orm import Session

from core.dependencies import get_db, get_current_user
from models.user import User
from schemas.document import (
    DocumentUploadResponse, 
    DocumentListItem, 
    DocumentStatusResponse
)
from services import document_service

router = APIRouter(tags=["Documents"])


@router.post("/upload", response_model=DocumentUploadResponse, status_code=status.HTTP_201_CREATED)
async def upload_document(
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
    file: UploadFile = File(...)
):
    """
    Uploads a document to Azure Blob Storage and creates a record in the database.
    """
    doc = document_service.upload_document(db, file=file, user_id=current_user.id)
    
    return {
        "document_id": doc.id,
        "filename": doc.filename,
        "status": doc.status,
        "message": "Document uploaded successfully and queued for processing."
    }


@router.get("/", response_model=List[DocumentListItem])
def list_documents(
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)]
):
    """
    Lists all documents belonging to the authenticated user.
    """
    return document_service.get_user_documents(db, user_id=current_user.id)


@router.get("/{document_id}", response_model=DocumentStatusResponse)
def get_document_status(
    document_id: str,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)]
):
    """
    Retrieves the status and metadata for a specific document.
    """
    doc = document_service.get_document_by_id(db, doc_id=document_id, user_id=current_user.id)
    if not doc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Document not found"
        )
    return doc


@router.get("/{document_id}/preview/markdown")
def preview_markdown(
    document_id: str,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)]
):
    """
    Returns the content of the extracted Markdown file.
    """
    content = document_service.get_document_preview(db, document_id, current_user.id)
    return {"content": content}


@router.get("/{document_id}/download/original")
def download_original(
    document_id: str,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)]
):
    """
    Downloads the original uploaded file.
    """
    content, filename, content_type = document_service.get_document_blob(
        db, document_id, current_user.id, is_markdown=False
    )
    return StreamingResponse(
        BytesIO(content),
        media_type=content_type,
        headers={"Content-Disposition": f"inline; filename={filename}"}
    )


@router.get("/{document_id}/download/markdown")
def download_markdown(
    document_id: str,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)]
):
    """
    Downloads the extracted Markdown file.
    """
    content, filename, content_type = document_service.get_document_blob(
        db, document_id, current_user.id, is_markdown=True
    )
    return StreamingResponse(
        BytesIO(content),
        media_type=content_type,
        headers={"Content-Disposition": f"inline; filename={filename}"}
    )

File: frontend/src/api/documents.js
import apiClient from './client';

export const uploadDocument = async (file) => {
  const formData = new FormData();
  formData.append('file', file);
  
  const response = await apiClient.post('/documents/upload', formData, {
    headers: {
      'Content-Type': 'multipart/form-data',
    },
  });
  return response.data;
};

export const listDocuments = async () => {
  const response = await apiClient.get('/documents/');
  return response.data;
};

export const getDocumentStatus = async (docId) => {
  const response = await apiClient.get(`/documents/${docId}`);
  return response.data;
};

export const getMarkdownPreview = async (docId) => {
  const response = await apiClient.get(`/documents/${docId}/preview/markdown`);
  return response.data;
};

export const getOriginalBlob = async (docId) => {
  const response = await apiClient.get(`/documents/${docId}/download/original`, {
    responseType: 'blob'
  });
  return response.data;
};

export const downloadOriginal = async (docId, filename) => {
  const response = await apiClient.get(`/documents/${docId}/download/original`, {
    responseType: 'blob'
  });
  const url = window.URL.createObjectURL(new Blob([response.data]));
  const link = document.createElement('a');
  link.href = url;
  link.setAttribute('download', filename);
  document.body.appendChild(link);
  link.click();
  link.remove();
};

export const downloadMarkdown = async (docId, filename) => {
  const response = await apiClient.get(`/documents/${docId}/download/markdown`, {
    responseType: 'blob'
  });
  const mdFilename = filename.split('.').slice(0, -1).join('.') + '.md';
  const url = window.URL.createObjectURL(new Blob([response.data]));
  const link = document.createElement('a');
  link.href = url;
  link.setAttribute('download', mdFilename);
  document.body.appendChild(link);
  link.click();
  link.remove();
};
